pupil_src/shared_modules/surface_tracker/surface.py

```diff
@@ -0,0 +1,513 @@
+"""
+(*)~---------------------------------------------------------------------------
+Pupil - eye tracking platform
+Copyright (C) 2012-2018 Pupil Labs
+
+Distributed under the terms of the GNU
+Lesser General Public License (LGPL v3.0).
+See COPYING and COPYING.LESSER for license details.
+---------------------------------------------------------------------------~(*)
+"""
+
+import numpy as np
+import cv2
+from abc import ABCMeta, abstractmethod
+import uuid
+
+import methods
+
+from surface_tracker import _Surface_Marker_Aggregate
+
+
+class Surface(metaclass=ABCMeta):
+    """A Surface is a quadrangel whose position is defined in relation to a set of
+    square markers in the real world. The markers are assumed to be in a fixed spatial
+    relationship and to be in plane with one another as well as the surface."""
+
+    def __init__(self, name="unknown", init_dict=None):
+        self.name = name
+        self.real_world_size = {"x": 1., "y": 1.}
+
+        # We store the surface state in two versions: once computed with the
+        # undistorted scene image and once with the still distorted scene image. The
+        # undistorted state is used to map gaze onto the surface, the distorted one
+        # is used for visualization. This is necessary because surface corners
+        # outside of the image can not be re-distorted for visualization correctly.
+        # Instead the slightly wrong but correct looking distorted version is
+        # used for visualization.
+        self.reg_markers_undist = {}
+        self.reg_markers_dist = {}
+        self.detected = False
+        self.img_to_surf_trans = None
+        self.surf_to_img_trans = None
+        self.dist_img_to_surf_trans = None
+        self.surf_to_dist_img_trans = None
+
+        self._required_obs_per_marker = 5
```

Comment:
I suspect that a few of those variable are not supposed to be changed (i.e. are constants). You could think about making them class variable with all-uppercase names, but actually I like it that similar variables are grouped together here, so I think it's ok to leave them as instance variables.